src/main/java/net/dv8tion/jda/api/requests/Route.java

```diff
@@ -432,68 +635,146 @@ public boolean equals(Object o)
-            this(baseRoute, compiledRoute, major, false);
+            this.baseRoute = original.baseRoute;
+            this.compiledRoute = original.compiledRoute;
+            this.major = original.major;
+            this.query = query;
         }
 
+        /**
+         * Returns a copy of this CompiledRoute with the provided parameters added as query.
+         * <br>This will use <a href="https://en.wikipedia.org/wiki/Percent-encoding" target="_blank">percent-encoding</a>
+         * for all provided <em>values</em> but not for the keys.
+         *
+         * <p><b>Example Usage</b><br>
+         * <pre>{@code
+         * Route.CompiledRoute history = Route.GET_MESSAGE_HISTORY.compile(channelId);
+         *
+         * // returns a new route
+         * route = history.withQueryParams(
+         *   "limit", 100
+         * );
+         * // adds another parameter ontop of limit
+         * route = route.withQueryParams(
+         *   "after", messageId
+         * );
+         *
+         * // now the route has both limit and after, you can also do this in one call:
+         * route = history.withQueryParams(
+         *   "limit", 100,
+         *   "after", messageId
+         * );
+         * }</pre>
+         *
+         * @param  params
+         *         The parameters to add as query, alternating key and value (see example)
+         *
+         * @throws IllegalArgumentException
+         *         If the number of arguments is not even or null is provided
+         *
+         * @return A copy of this CompiledRoute with the provided parameters added as query
+         */
         @Nonnull
         @CheckReturnValue
-        public CompiledRoute withQueryParams(String... params)
+        public CompiledRoute withQueryParams(@Nonnull String... params)
         {
-            Checks.check(params.length >= 2, "params length must be at least 2");
-            Checks.check(params.length % 2 == 0, "params length must be a multiple of 2");
+            Checks.notNull(params, "Params");
+            Checks.check(params.length >= 2, "Params length must be at least 2");
+            Checks.check((params.length & 1) == 0, "Params length must be a multiple of 2");
 
-            StringBuilder newRoute = new StringBuilder(compiledRoute);
+            List<String> newQuery;
+            if (query == null)
+            {
+                newQuery = new ArrayList<>(params.length / 2);
+            }
+            else
+            {
+                newQuery = new ArrayList<>(query.size() + params.length / 2);
+                newQuery.addAll(query);
+            }
 
-            for (int i = 0; i < params.length; i++)
-                newRoute.append(!hasQueryParams && i == 0 ? '?' : '&').append(params[i]).append('=').append(params[++i]);
+            // Assuming names don't need encoding
+            for (int i = 0; i < params.length; i += 2)
+                newQuery.add(params[i] + '=' + EncodingUtil.encodeUTF8(params[i + 1]));
```

Comment:
As a note, query strings with no value like `?use_flag` are a thing. We might want to check if the provided value for a key is `""` and not add the `=` for it.

Reply:
Can also cross that bridge when we come to it, but thought I'd mention it

Reply:
Usually It's valid to do something like `?use_flag=` with no value attached as well.